cmd/gateway-main.go

```diff
@@ -277,7 +280,12 @@ func gcsGatewayMain(ctx *cli.Context) {
 		cli.ShowCommandHelpAndExit(ctx, "gcs", 1)
 	}
 
-	if !isValidGCSProjectIDFormat(ctx.Args().First()) {
+	projectID := ctx.Args().First()
+	if projectID == "" && os.Getenv("GOOGLE_APPLICATION_CREDENTIALS") == "" {
+		errorIf(errGCSProjectIDNotFound, "project-id should be provided as argument or GOOGLE_APPLICATION_CREDENTIALS should be set with path to credentials.json")
+		cli.ShowCommandHelpAndExit(ctx, "gcs", 1)
+	}
+	if projectID != "" && !isValidGCSProjectIDFormat(projectID) {
```

Comment:
As project ID from command line has higher precedence over env var `GOOGLE_APPLICATION_CREDENTIALS`.   I think below `if` block would be clear

```go
credsJSONFile := os.Getenv("GOOGLE_APPLICATION_CREDENTIALS")
if projectID == "" {
    if credsJSONFIle == "" {
       fatalIf(...)
    }
    if projectID, err = gcsParseProjectID(credsJSONFile); err != nil {
        fatalIf(err, ...)
    }
}
```

This way `newGCSGateway()` could use `projectID` argument without knowing about env var `GOOGLE_APPLICATION_CREDENTIALS`

Reply:
I tried to do this @balamurugana but our source code structure does not allow this flow, it needs a refactor to allow it.

i.e I tried to make the flow such that newGCSGateway() does not access GOOGLE_APPLICATION_CREDENTIALS but it was not possible. It needs a separate refactor of the functions in gateway-main.go

Reply:
OK. Probably we need to open an issue?